src/libraries/illink-sharedframework.targets

```diff
@@ -0,0 +1,72 @@
+<Project>
+
+  <Target Name="ILLinkTrimSharedFramework"
+          AfterTargets="Build"
+          DependsOnTargets="SetCommonILLinkArgs">
+
+    <PropertyGroup>
+      <!-- default action for core assemblies -->
+      <ILLinkArgs>$(ILLinkArgs) -c link</ILLinkArgs>
+      <!-- update debug symbols -->
+      <ILLinkArgs>$(ILLinkArgs) -b true</ILLinkArgs>
+      <!-- suppress warnings with the following codes:
+           IL2006: The generic parameter 'T' from A with dynamically accessed member kinds B is passed into the generic parameter
+                   'T' from 'System.Lazy<T>' which requires dynamically accessed member kinds 'PublicParameterlessConstructor'
+           IL2009: Could not find method A in type B specified in resource C
+           IL2025: Duplicate preserve of A in B
+           IL2026: Calling A which has B can break functionality when trimming application code. The target method might be removed.
+           IL2035: Unresolved assembly A in DynamicDependencyAttribute on B
+           IL2050: P/invoke method A declares a parameter with COM marshalling. Correctness of COM interop
+                   cannot be guaranteed after trimming. Interfaces and interface members might be removed.
+      -->
+      <ILLinkArgs>$(ILLinkArgs) --nowarn IL2006;IL2009;IL2025;IL2026;IL2035;IL2050</ILLinkArgs>
+    </PropertyGroup>
+
+     <!-- Retrieve CoreLib's targetpath via GetTargetPath as it isn't binplaced yet. -->
```

Comment:
I'm not sure I understand this. Why can't we use `$(MicrosoftNetCoreAppRuntimePackNativeDir)` instead?

Reply:
Because `$(MicrosoftNetCoreAppRuntimePackNativeDir` is populated in the [pretest project](https://github.com/dotnet/runtime/blob/809a06f45161ae686a06b9e9ccc2f45097b91657/src/libraries/pretest.proj#L96) (which occurs after the execution of this step) and it doesn't seem right to add this target (`ILLinkTrimSharedFramework`) to that project.

Reply:
My naive thinking is that there should be somewhere in the build (that doesn't rely on tests) that we can hook in where the entire runtimepack is built. But I'm not seeing where that place is. Maybe @safern @ViktorHofer @joperezr or @ericstj might know.

But for now, what you have works so let's go with that.

Reply:
SPC and the runtime gets binplaced into libraries runtime pack and testhost as part of runtime.depproj. 

Runtime.depproj is built as part of pretest.proj which is the last step of the `libs` subset. Which happens after `libs.src` that is the subset that builds `src.proj`. That is why `wasm.targets` also follows that approach.

FWIW we could introduce a common target to get SPC target path instead of duplicating the logic to call MSBuild on it with GetTargetPath target.

Reply:
> FWIW we could introduce a common target to get SPC target path instead of duplicating the logic to call MSBuild on it with GetTargetPath target.

It feels to me that the correct fix here would be to fully populate the runtimepack earlier, and not make it dependent on testing.

Reply:
The reason why having the runtime binplace at pretest is handy, is for inner loop development. For example if I update code in SPC after a vertical build, in order to run tests against the new SPC, I don't have to do a full libraries vertical, I can just do: `build libs.pretest -rc <runtimeConfig>` and we actually use this in CI to avoid building a vertical against each runtime we run tests (mono, checked coreclr, and release coreclr) and just replace the runtime with the one we want to run tests against by building the pretest subset.

Reply:
> It feels to me that the correct fix here would be to fully populate the runtimepack earlier, and not make it dependent on testing.

I agree with @eerhardt, we have talked about this couple times already. I believe we should binplace the runtime artifacts incrementally after they are built (CoreLib, coreclr.dll, etc). For CI we just need a target that re-binplaces the runtime artifacts (to preserve the existing behavior which allows to build coreclr and libraries in parallel).

> The reason why having the runtime binplace at pretest is handy, is for inner loop development. For example if I update code in SPC after a vertical build, in order to run tests against the new SPC, I don't have to do a full libraries vertical, I can just do: build libs.pretest -rc <runtimeConfig>

I don't understand why binplacing the runtime artifacts in pretest is handy for inner loop development instead of binplacing the runtime assets at the time they are (re-)built, which doesn't require a libraries rebuild.

Reply:
> I don't understand why binplacing the runtime artifacts in pretest is handy for inner loop development instead of binplacing the runtime assets at the time they are (re-)built, which doesn't require a libraries rebuild.

We've also talked about this a couple of times. If you build for example System.Runtime which has a P2P to SPC, and don't specify RuntimeConfiguration as an argument you would potentially be binplacing a `Debug` build of SPC when probably you built the runtime for release, this could cause a mismatch in between the native bits and the managed bits of coreclr which is not recommended:
https://github.com/dotnet/runtime/pull/38885#issuecomment-655082212

So we should be careful on how we do that.

Reply:
Btw, I agree we should change this if possible, I'm just saying we should be really careful how and where we put this binplacing to happen.

Reply:
> We've also talked about this a couple of times. If you build for example System.Runtime which has a P2P to SPC, and don't specify RuntimeConfiguration as an argument you would potentially be binplacing a Debug build of SPC when probably you built the runtime for release, this could cause a mismatch in between the native bits and the managed bits of coreclr which is not recommended:
#38885 (comment)
Btw, I agree we should change this if possible, I'm just saying we should be really careful how and where we put this binplacing to happen.

I think we all agree that we should change when we binplace the runtime artifacts and I agree that it should be done with the highest diligence. We should do this as part of https://github.com/dotnet/runtime/issues/38034. 

> Also, I wouldn't like removing the build.pretest binplacing the runtime as it is handy for CI as well as I explained above... if we remove that, then provide another hook to binplace the runtime as an individual step.

That's what I meant with: 
> For CI we just need a target that re-binplaces the runtime artifacts (to preserve the existing behavior which allows to build coreclr and libraries in parallel).

Reply:
> That's what I meant with:

I somehow missed that part 🤦

> We should do this as part of #38034.

Sounds good.